provisiond: implement cryptography primitives for resource identification

zaibon opened the pull request:
fixes #83

zaibon committed:
provisiond: implement cryptography primitives for resousrce identification

fixes #83

zaibon committed:
networkd: set the public iface  type during public iface registration

zaibon committed:
provision: rename Reservation.Tenant to Reservation.User

zaibon committed:
provisiond: make the containers namespace and ID determinictic

zaibon committed:
provisiond: make volume name deterministic

zaibon committed:
provisiond: make the user ID a string instead of a identity.Identifier

makes it easier to unmarshall from the network

zaibon committed:
provisiond: verfiy the signature of the reservation before processing it

zaibon committed:
tfuser: allow to configure provision URL using environment variable

zaibon committed:
tfuser: add command to generate an identity

muhamadazmy commented on modules/provision/container.go:
```diff
@@ -93,10 +92,15 @@ func ContainerProvision(ctx context.Context, reservation Reservation) (interface
+		containerID,
 		modules.Container{
-			Name:   uuid.New().String(),
+			Name:   reservation.User,
```
This can cause conflict if the same user is hosting multiple containers on the same node. I suggest using the containerID.

zaibon replied:
yep this is mistake, I inverted container tenant namespace and container name

muhamadazmy commented on modules/provision/crypto.go:
```diff
@@ -0,0 +1,101 @@
+}
+
+// Verify verifies the signature of the reservation
+func Verify(r Reservation) error {
```
Don't see a reason why this can't be an instance method.

```suggestion
func (r *Reservation) Verify() error {
```

zaibon replied:
was just to not overload the API of the Reservation struct. could be change to a method

muhamadazmy commented on modules/provision/crypto.go:
```diff
@@ -0,0 +1,101 @@
+	return crypto.Verify(publicKey, buf.Bytes(), r.Signature)
+}
+
+func Hash(r Reservation) ([]byte, error) {
```
Same here.

```suggestion
func (r *Reservation) Hash() ([]byte, error) {
```

muhamadazmy commented on modules/provision/provision.go:
```diff
@@ -27,26 +27,19 @@ const (
-	// we need to report to once the reservation is done
-	ReplyTo ReplyTo `json:"reply-to"`
+	// Identification of the user requesting the reservation
+	User string `json:"user_id"`
```
Why the `reply-to` was dropped? we still need to notify the user `3bot` on the reservation result. Do u have another idea to do this?

zaibon replied:
This is not an information that needs to be in the reservation. We already have the userID (3bot) from which you get the current address of this 3bot. This address might change so no point having it in the reservation

muhamadazmy reviewed (commented):
Except for few comments, the PR looks good. But there is one thing i don't get.

The signature now verify that a reservation is authentic and hasn't been tempered with. But it still doesn't verify that a resource ID pointed to by reservation is actually owned by the user. For example a VolumeID in a container reservation. 

Also, the volumeID now does not carry any information about the volume, hence the container provision logic is completely wrong now, assuming that the volumeID is the source of a mount. 

So to sum up and rephrase:
- The idea of using a recomputable hash as an ID is good.
- The node can't use the ID to point to a reservation, hence it can't figure out what u pointing to (volumeID to actual volume path) unless the node start tracking all active reservation locally in some sort of dict.
- The node can't verify that you are the owner of a resource that you are referring to (again the container -> volume example) unless again the node can track (or at least be able to query) all active reservations.
- Current code assumes that volume id is the same as the volume path which is wrong.

zaibon commented:
> The signature now verify that a reservation is authentic and hasn't been tempered with. But it still doesn't verify that a resource ID pointed to by reservation is actually owned by the user. For example a VolumeID in a container reservation.

This is correct this is not in the PR yet. keep in mind this is still a draft and more things needs to be added.

> The node can't use the ID to point to a reservation, hence it can't figure out what u pointing to (volumeID to actual volume path) unless the node start tracking all active reservation locally in some sort of dict.

This is exactly the idea, see https://github.com/threefoldtech/zosv2/issues/83#issuecomment-515235594 for more detail

zaibon commented:
Ok I took time to read you again @muhamadazmy and all your point are valid.
The way I wanted to generated new workload ID based on the reservation won't work, cause I then create an indirection between the reservation ID and the workload ID that I cannot resolve without storing the mapping somewhere (we don't want state in the node)

So really the only solution I have is to directly use the reservation ID as workload ID, so the mapping is direct and doesn't requires to store any stage. 
I just need to double check how are the reservation ID will be generated (I think BCDB will be the done doing that)

zaibon committed:
tfuser: improve network schema generation and workload provisioning

- generate and print wireguard private key when adding a user to a
network
- allow to generate a wg-quick configuration for a specific user from a
network object
- use the user private ket to sign the reservation schema

zaibon committed:
provision: change bcdb provision logic

- the bcdb now allows to poll for new reservation or list all
reservation for a specific node
- the bcdb is the one setting the ID on a reservation

zaibon committed:
storage: add a Path method on the storage interface

this is used to get the mountpoint of a subvolume from its name

zaibon committed:
provision: check the ownership of a volume when mounting it inside a
container

When creating a container, we now check that any volume in the mounts
list also belong to the same user. This to prevent any attacker to mount
a volume from another user inside its own container and tamper with the
data.

zaibon committed:
provisiond: try to provision existing workload at boot

Provisiond now has a bootstrap step when it starts.
It will get all the reservation for the node, and try to provision all
of them.

Once this is done, it starts the provision engine and start watching for
new reservations

zaibon committed:
go mod tidy all binaries

zaibon commented:
In my last batch of commits I updated the tnodb_mock so he's the one generating reservation ID. This is going to map better with own bcdb works. The generated id now looks like `r-$id` where `$id` is an incremental int. We can change it later on. 

But one thing to keep in mind is this id needs to be relatively short, this is because since we use the reservation ID as container name, the ID cannot be too long (not sure how much) cause containerd create a unixsocket with the tenant namespace+container name
the tenant namespace being the user ID, it is already 44 chars long
Most probably the same kind of limitation applies for the volumes name in btrfs, which is now also the reservation ID.

The rest of the commits are some clean up and update of tfuser to map the new flow. 

This should be already big enough like this. I'll open this PR for review, let's see if we can merge, then let's play a bit with that built to gather info on what to do next.

zaibon committed:
update documentation about basic network creation

muhamadazmy commented:
> The generated id now looks like r-$id where $id is an incremental int. We can change it later on.

Incremental IDs in a distributed ecosystem, this gonna be really hard or almost impossible to implement

zaibon commented:
> Incremental IDs in a distributed ecosystem, this gonna be really hard or almost impossible to implement

As far as I know this is how BCDB works. But like I said, this is implement like this to allow us to make progress now. We'll have to adapt once BCDB is live.

muhamadazmy commented on cmds/upgraded/main.go:
```diff
@@ -4,6 +4,8 @@ import (
 	"os"
 	"time"
 
+	"github.com/threefoldtech/zosv2/modules/zinit"
```
I don't believe this entire file should be part of this Pull Request.

zaibon replied:
yep indeed, I moved it to https://github.com/threefoldtech/zosv2/pull/155
